app/core/models.py

```diff
@@ -44,6 +45,55 @@ class Meta:
         get_latest_by = ("created_at",)
 
 
+def _timestamped_log(log):
+    return " ".join([time.strftime("[%c %Z]", time.gmtime()), log.decode("utf-8")])
+
+
+class Logged(Timestamped, StatusMixin):
+    log_stdout = models.TextField(blank=True)
+    log_stderr = models.TextField(blank=True)
+
+    log_handler = None
+
+    class Meta:
+        abstract = True
+
+    class LogHandler:
+        def __init__(self, instance):
+            self.instance_id = instance.pk
+            self.cls = type(instance)
+
+        @property
+        def _update(self):
+            return self.cls.objects.filter(pk=self.instance_id).update
+
+        def handle_stdout(self, log):
+            self._update(
+                log_stdout=Concat(
+                    models.F("log_stdout"), models.Value(_timestamped_log(log))
+                )
+            )
+
+        def handle_stderr(self, log):
+            self._update(
+                log_stderr=Concat(
+                    models.F("log_stderr"), models.Value(_timestamped_log(log))
+                )
+            )
+
+    def append(self, stdout=None, stderr=None):
+        update_fields = []
+        if stdout is not None:
+            self.log_stdout = Concat(models.F("log_stdout"), models.Value(stdout))
+            update_fields.append("log_stdout")
+        if stderr is not None:
+            self.log_stderr = Concat(models.F("log_stderr"), models.Value(stderr))
+            update_fields.append("log_stderr")
+
+        if update_fields:
+            self.save(update_fields=update_fields)
```

Comment:
Was the call to save what we were missing last time?

Reply:
no, last time we were calling save without update_fields on an object that didn't have the logs updated. So the in-memory version of the log fields were out of date, and then that's what was written to the db